lib/PreviewProvider.js

```
@@ -0,0 +1,226 @@
+      site.sanitizedURL = this._sanitizeURL(site.url);
+      site.cacheKey = this._createCacheKey(site.url);
+      if (!dedupedSites.has(site.cacheKey)) {
+        dedupedSites.set(site.cacheKey, site);
+      }
+    });
+    return Array.from(dedupedSites.values());
+  },
+
+  /**
+    * Filter cached links out, and store their embedly data
+    */
+  getCachedLinks(links) {
+    let cachedLinks = [];
+    let sites = this._uniqueLinks(links.filter(this._URLFilter(URL_FILTERS)));
+    sites.forEach(link => {
+      if (link && ss.storage.embedlyData[link.cacheKey] && link.cacheKey === ss.storage.embedlyData[link.cacheKey].cacheKey) {
+        cachedLinks.push(ss.storage.embedlyData[link.cacheKey]);
+      }
+    });
+
+    return cachedLinks;
+  },
+
+  /**
+   * Filter out new links and request them from embedly
+   */
+  saveNewLinks: Task.async(function*(links) {
+    let sites = this._uniqueLinks(links.filter(this._URLFilter(URL_FILTERS)));
+    let newLinks = sites.filter(link => link && !ss.storage.embedlyData[link.cacheKey]);
+    let requestQueue = [];
+    while (newLinks.length !== 0) {
+      // we have some new links we need to fetch the embedly data for, put them on the queue
+      requestQueue.push(newLinks.splice(0, EMBEDLY_PROXY_MAX_LINKS));
+    }
+    // for each bundle of 25 links, create a new request to embedly
+    requestQueue.forEach(requestBundle => {
+      this._fetchAndCache(requestBundle);
+    });
+  }),
+
+  /**
+   * Makes the necessary requests to embedly to get data and store it in the cache
+   */
+  _fetchAndCache: Task.async(function*(newLinks) {
+    let json = null;
+    // extract only the sanitized link urls to send to embedly
+    let linkURLs = newLinks.map(link => link.sanitizedURL);
+    try {
+      let response = yield fetch(EMBEDLY_ENDPOINT, {
+        method: "POST",
+        body: JSON.stringify({urls: linkURLs}),
+        headers: {"Content-Type": "application/json"}
+      });
+      if (response.ok) {
```

What happens (or should happen) if we get a response from the embedly-proxy and the response is NOT `ok`?
Do we need an `else { ... }` here, or am I missing something? I just wanted to make sure there isn't some weird case where the proxy returns an error but something isn't thrown/caught and we're just silently ignoring something.